ui/src/components/setlist/SetlistSongTable.tsx

```diff
@@ -29,156 +29,185 @@ import SetlistChangeKeyModal from './SetlistChangeKeyModal';
 
 interface SetlistSongsTableProps {
   songList: SongSetlistSchema[];
-  setSongList: React.Dispatch<React.SetStateAction<SongSetlistSchema[]>>;
+  setSongList?: React.Dispatch<React.SetStateAction<SongSetlistSchema[]>>;
+  readOnly?: boolean;
 }
 
-const SetlistSongsTable: React.FC<SetlistSongsTableProps> = ({ songList, setSongList }) => {
-  const [sortedSongList, setSortedSongList] = useState<SongSetlistSchema[]>(songList);
+const SetlistSongsTable: React.FC<SetlistSongsTableProps> = ({
+  songList,
+  setSongList,
+  readOnly = false,
+}) => {
+  // State management
+  const [sortedSongList, setSortedSongList] = useState<SongSetlistSchema[]>([]);
+  const [menuState, setMenuState] = useState({
+    anchorEl: null as HTMLElement | null,
+    currentSongId: null as string | null,
+  });
 
+  // Modal states with non-null song when open
+  const [modals, setModals] = useState({
+    preview: {
+      open: false,
+      song: null as SongSetlistSchema | null,
+    },
+    changeKey: {
+      open: false,
+      song: null as SongSetlistSchema | null,
+    },
+  });
+
```

Comment:
[important]: setting the modal type like this as a field of an object is quite dangerous, its better if you make a whole new object type, lets say
```
  const enum ModalType {
    ModalPreview = 'preview',
    ModalChangeKey = 'changeKey',
  }
  type ModalTable = {
    modalType: ModalType;
    open: boolean;
    song: SongSetlistSchema | null;
  };
```
and use this object type as the state type.